cisco_duo.trust_monitor: Fix 400 and 401 errors in Trust Monitor

kcreddy opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## Proposed commit message
[cisco_duo] Fix `400` and `401` errors in Trust Monitor.

[Cisco API documentation](https://duo.com/docs/adminapi#authentication) mentions that the `Date` header should equal current timestamp. 
In current CEL program, the previous execution's `state.date` is being captured and 
used in pagination and subsequent requests leading to possible `40103` errors. 
Also, although the API doc doesn't provide the information, users observed error
`400`: `mintime must be within the past 180 days`.

This PR:
- Fixes `40103` by creating `Date` header at each CEL request preparation.
- Fixes `400` by checking `duration` between `mintime` and `maxtime` doesn't cross
  `180d`.

-----

Note to reviewers: `dyn(state)` was used in few occasions in CEL program to avoid errors namely: 
- `type 'string' does not support field selection`
- `no such overload string(double)`
See https://github.com/elastic/integrations/pull/11621#discussion_r1827378901 and https://github.com/elastic/integrations/pull/11621#discussion_r1829191166

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [ ] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

**Current tracer logs, 1st pagination request:**
The `timestamp` and `Date` header are not same.
`{"log.level":"debug","@timestamp":"2024-11-05T11:14:07.296Z","message":"HTTP request",.........\r\nDate: Tue, 05 Nov 2024 11:14:05 +0000\r\nAccept......}`

**After the change: tracer logs, 1st pagination request:**
The `timestamp` and `Date` header are same.
`{"log.level":"debug","@timestamp":"2024-11-05T11:05:47.546Z","message":"HTTP request","......\r\nDate: Tue, 05 Nov 2024 11:05:47 +0000\r\nAccept......}`

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/17835)
* Commit: ef23a21776ec387d06267fce7ab56fff2bac4301

### Failed CI Steps
* [Check integrations cisco_duo](https://buildkite.com/elastic/integrations/builds/17835#0192f348-7bbb-4b93-8f25-6d6ba9852a3f)

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17835","number":17835,"commit":"ef23a21776ec387d06267fce7ab56fff2bac4301"}],"number":17835}
buildkite-pr-comment-->

kcreddy commented on packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs:
```diff
@@ -28,120 +28,110 @@ program: |
+                            ?"last_published": (has(body.?response.events) && size(body.response.events) > 0) ?
+                                optional.of(
+                                    body.response.events.map(e, e.surfaced_timestamp).max().as(last_timestamp,
+                                        !has(dyn(state).?cursor.last_published) ?
```
`dyn` added in few places to avoid error: `type 'string' does not support field selection`

kcreddy commented on packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs:
```diff
@@ -28,120 +28,110 @@ program: |
+            "url_base": state.url.trim_right("/"),
+            "url_path": "/admin/v1/trust_monitor/events",
+            "query_string": {
+                 "limit": [string(dyn(state).limit)],
```
`dyn(state)` added here to avoid error: `no such overload string(double)`

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/17914)
* Commit: 25817536a612e9e60c222139a62d5e380740d2d6

### History
* :broken_heart: [Build #17835](https://buildkite.com/elastic/integrations/builds/17835) failed ef23a21776ec387d06267fce7ab56fff2bac4301

cc @kcreddy

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17914","number":17914,"commit":"25817536a612e9e60c222139a62d5e380740d2d6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17835","number":17835,"commit":"ef23a21776ec387d06267fce7ab56fff2bac4301"}],"number":17914}
buildkite-pr-comment-->

kcreddy commented:
Addressed comments: https://github.com/elastic/integrations/pull/11640#discussion_r1830739341 and https://github.com/elastic/integrations/pull/11640#discussion_r1830743284 and also @chemamartinez's suggestion on using `now()` vs `now` in `date` calculation. 

The change is in commit [edcac5a](https://github.com/elastic/integrations/pull/11621/commits/edcac5aa8461a94a1349be7e44f9cbb133db6572)

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/17970)
* Commit: edcac5aa8461a94a1349be7e44f9cbb133db6572

### History
* :green_heart: [Build #17914](https://buildkite.com/elastic/integrations/builds/17914) succeeded 25817536a612e9e60c222139a62d5e380740d2d6
* :broken_heart: [Build #17835](https://buildkite.com/elastic/integrations/builds/17835) failed ef23a21776ec387d06267fce7ab56fff2bac4301

cc @kcreddy

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17970","number":17970,"commit":"edcac5aa8461a94a1349be7e44f9cbb133db6572"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17914","number":17914,"commit":"25817536a612e9e60c222139a62d5e380740d2d6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17835","number":17835,"commit":"ef23a21776ec387d06267fce7ab56fff2bac4301"}],"number":17970}
buildkite-pr-comment-->

kcreddy commented on packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs:
```diff
@@ -34,10 +34,7 @@ program: |
-                    string(mx * 1000)
-                : 
-                    string(oneeigthydays * 1000)
+                mx < oneeigthydays ? (state.want_more ? string(state.maxtime) : string(mx * 1000)) : string(oneeigthydays * 1000)
```
This is not formatted due to following error (if formatted):

`failed to check program: failed compilation: ERROR: <input>:-1:0: error recovery token lookahead limit exceeded: 256\nERROR: <input>:13:33: Syntax error: mismatched input '?' expecting ':'\n |                 state.want_more ? \n | ................................^\nERROR: <input>:17:13: Syntax error: mismatched input ':' expecting ')'\n |             : \n | ............^\nERROR: <input>:18:44: Syntax error: mismatched input ')' expecting {'}', ','}\n |                 string(oneeigthydays * 1000)\n | ...........................................^\nERROR: <input>:20:5: Syntax error: mismatched input ')' expecting <EOF>\n |     )\n | ....^ accessing config`

kcreddy replied:
This is not formatted because of following error (if formatted):

`failed to check program: failed compilation: ERROR: <input>:-1:0: error recovery token lookahead limit exceeded: 256\nERROR: <input>:13:33: Syntax error: mismatched input '?' expecting ':'\n |                 state.want_more ? \n | ................................^\nERROR: <input>:17:13: Syntax error: mismatched input ':' expecting ')'\n |             : \n | ............^\nERROR: <input>:18:44: Syntax error: mismatched input ')' expecting {'}', ','}\n |                 string(oneeigthydays * 1000)\n | ...........................................^\nERROR: <input>:20:5: Syntax error: mismatched input ')' expecting <EOF>\n |     )\n | ....^ accessing config`

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/17971)
* Commit: 1a39f3e2c4cfa12ec177556582af494e0b50b4bc

### History
* :green_heart: [Build #17970](https://buildkite.com/elastic/integrations/builds/17970) succeeded edcac5aa8461a94a1349be7e44f9cbb133db6572
* :green_heart: [Build #17914](https://buildkite.com/elastic/integrations/builds/17914) succeeded 25817536a612e9e60c222139a62d5e380740d2d6
* :broken_heart: [Build #17835](https://buildkite.com/elastic/integrations/builds/17835) failed ef23a21776ec387d06267fce7ab56fff2bac4301

cc @kcreddy

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17971","number":17971,"commit":"1a39f3e2c4cfa12ec177556582af494e0b50b4bc"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17970","number":17970,"commit":"edcac5aa8461a94a1349be7e44f9cbb133db6572"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17914","number":17914,"commit":"25817536a612e9e60c222139a62d5e380740d2d6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17835","number":17835,"commit":"ef23a21776ec387d06267fce7ab56fff2bac4301"}],"number":17971}
buildkite-pr-comment-->

chrisberkhout commented on packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs:
```diff
@@ -28,120 +28,112 @@ program: |
-                "mintime": state.?cursor.last_published.orValue(int(now - duration(state.initial_interval)) * 1000),
-                "maxtime": int(now - duration("2m")) * 1000,
-                "date": now.format(time_layout.RFC1123Z),
+                "mintime": string(state.?cursor.last_published.orValue(int(now - duration(state.initial_interval)) * 1000)),
```
```suggestion
                "mintime": string(int(state.?cursor.last_published.orValue(int(now - duration(state.initial_interval)) * 1000))),
```

I think this outer `int` is needed for the case that `state.cursor.last_published` is used.

Large int values will be formatted in floating point notation if not explicitly cast to int. It doesn't just happen with JSON input data, it happens with int values passed forward in state.

For example:
```
mito -max_executions 3 -data <(echo '{"t": 1746522206000}') <(echo '{
  "t": state.t,
  "int_t": int(state.t),
  "str_t": string(state.t),
  "str_int_t": string(int(state.t)),
  "want_more": true,
}')
```
```
{
	"int_t": 1746522206000,
	"str_int_t": "1746522206000",
	"str_t": "1.746522206e+12",
	"t": 1746522206000,
	"want_more": true
}
{
	"int_t": 1746522206000,
	"str_int_t": "1746522206000",
	"str_t": "1.746522206e+12",
	"t": 1746522206000,
	"want_more": true
}
{
	"int_t": 1746522206000,
	"str_int_t": "1746522206000",
	"str_t": "1.746522206e+12",
	"t": 1746522206000,
	"want_more": true
}
```

kcreddy replied:
This is nice. I had same errors, didn't know where `double` was coming from. Guess I could remove one of the `dyn(`

kcreddy replied:
Guess not, removing that `dyn` still leads to same error `no such overload string(double)`. So retained the `dyn`.

Updated as per suggestion in [9122974](https://github.com/elastic/integrations/pull/11621/commits/912297463ec88b9f1b1f22bbf47561b62f8ed539)

chrisberkhout commented on packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs:
```diff
@@ -28,120 +28,112 @@ program: |
-            {
+    ).as(state, state.with({
+        // the duration between maxtime and mintime should be less than 180d to avoid error_code 400 from API.
+        "maxtime": int(now - duration("2m")).as(mx, 
+            (int(timestamp(int(state.mintime) / 1000) + duration("4320h") - duration("2m"))).as(oneeigthydays, 
+                mx < oneeigthydays ? (state.want_more ? string(state.maxtime) : string(mx * 1000)) : string(oneeigthydays * 1000)
```
```suggestion
        "maxtime": [int(timestamp(int(state.mintime)/1000) + duration("4320h")), int(now - duration("2m"))].min()
```

Easier to read:
Go 180 days forward, or up to 2 mins ago, whichever is less.

kcreddy replied:
But this doesn't check the `state.want_more ?` case which is required to pass same `maxtime` to the pagination requests and not re-calculated one.

chrisberkhout replied:
Sorry, I missed that.

In the original, if `state.want_more == true` it will not reuse the existing `state.maxtime` if a newly calculated value crosses the current time.

I updated the suggestion to check `state.want_more` and only calculate a new value if it's an initial page.

Is that correct?

Feel free to take whatever parts help.

chrisberkhout commented on packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs:
```diff
@@ -28,120 +28,112 @@ program: |
+                 "limit": [string(dyn(state).limit)],
+                 "maxtime": [string(state.maxtime)],
+                 "mintime": [string(state.mintime)],
+                 ?"offset": has(state.next_offset) ?
+                    optional.of([string(state.next_offset)])
+                :
+                    optional.none(),
```
```suggestion
                 ?"offset": state.?next_offset.optMap(v, [string(v)]),
```

kcreddy replied:
Updated as suggested.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/18011)
* Commit: 912297463ec88b9f1b1f22bbf47561b62f8ed539

### History
* :green_heart: [Build #17971](https://buildkite.com/elastic/integrations/builds/17971) succeeded 1a39f3e2c4cfa12ec177556582af494e0b50b4bc
* :green_heart: [Build #17970](https://buildkite.com/elastic/integrations/builds/17970) succeeded edcac5aa8461a94a1349be7e44f9cbb133db6572
* :green_heart: [Build #17914](https://buildkite.com/elastic/integrations/builds/17914) succeeded 25817536a612e9e60c222139a62d5e380740d2d6
* :broken_heart: [Build #17835](https://buildkite.com/elastic/integrations/builds/17835) failed ef23a21776ec387d06267fce7ab56fff2bac4301

cc @kcreddy

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18011","number":18011,"commit":"912297463ec88b9f1b1f22bbf47561b62f8ed539"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17971","number":17971,"commit":"1a39f3e2c4cfa12ec177556582af494e0b50b4bc"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17970","number":17970,"commit":"edcac5aa8461a94a1349be7e44f9cbb133db6572"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17914","number":17914,"commit":"25817536a612e9e60c222139a62d5e380740d2d6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17835","number":17835,"commit":"ef23a21776ec387d06267fce7ab56fff2bac4301"}],"number":18011}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/18014)
* Commit: 0a796ab4d6f26d5682e5c8a26163dc2154dce7a9

### History
* :green_heart: [Build #18011](https://buildkite.com/elastic/integrations/builds/18011) succeeded 912297463ec88b9f1b1f22bbf47561b62f8ed539
* :green_heart: [Build #17971](https://buildkite.com/elastic/integrations/builds/17971) succeeded 1a39f3e2c4cfa12ec177556582af494e0b50b4bc
* :green_heart: [Build #17970](https://buildkite.com/elastic/integrations/builds/17970) succeeded edcac5aa8461a94a1349be7e44f9cbb133db6572
* :green_heart: [Build #17914](https://buildkite.com/elastic/integrations/builds/17914) succeeded 25817536a612e9e60c222139a62d5e380740d2d6
* :broken_heart: [Build #17835](https://buildkite.com/elastic/integrations/builds/17835) failed ef23a21776ec387d06267fce7ab56fff2bac4301

cc @kcreddy

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18014","number":18014,"commit":"0a796ab4d6f26d5682e5c8a26163dc2154dce7a9"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18011","number":18011,"commit":"912297463ec88b9f1b1f22bbf47561b62f8ed539"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17971","number":17971,"commit":"1a39f3e2c4cfa12ec177556582af494e0b50b4bc"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17970","number":17970,"commit":"edcac5aa8461a94a1349be7e44f9cbb133db6572"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17914","number":17914,"commit":"25817536a612e9e60c222139a62d5e380740d2d6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/17835","number":17835,"commit":"ef23a21776ec387d06267fce7ab56fff2bac4301"}],"number":18014}
buildkite-pr-comment-->

chrisberkhout reviewed (approved):
In this PR, since there was some refactoring with indentation changes, it would have been helpful to separate changes in that category from the functional changes, just to be able to check things in smaller diffs.

Looks like good and correct changes. :+1:
